### PR TITLE
remove dependency to maskpasswords

### PR DESCRIPTION
**Pull request by mwinter69**
by changing to OutputStream saving a freestyle project will no longer fail

### CODE REVIEW

**Comment by jakub-bochenski** [comment]
Is that really all? I think I saw some other references to mask-passwords classes in the code.

Have you verified this change on a Jenkins instance w/o mask-passwords?

**Comment by mwinter69** [comment]
actually it doens't work. Seems in my test I still had it installed somehow.

**Comment by mwinter69** [comment]
Potentially we can completely omit the maskpasswords stuff. The BuildWrapper of MaskPasswords has an ordinal of 1000 meaning it will be most likely the last outputstream that is initialized and thus will be the first that is called for each log line. Why should we filter what is already filtered. Additionally maskpasswords has the new feature of regex which is not yet reflected in the code here.

**Comment by mwinter69** [comment]
It is possible when using a higher ordinal that passwords are masked. Order of the buildwrappers in the config.xml is the key.
It will not work when you manually create the config.xml of a job and put the Maskpasswords before the logstash build wrapper. 
But when configuring this via the UI it works fine.
latest commit does this

**Comment by jakub-bochenski** [comment]
What about https://github.com/jenkinsci/logstash-plugin/blob/9046206c03c8c76ab9fec403000376b6751698a8/src/main/java/jenkins/plugins/logstash/LogstashBuildWrapper.java#L84 ?

> It will not work when you manually create the config.xml of a job and put the Maskpasswords before the logstash build wrapper.

What about pipeline? I expect one would need to remember to declare the maskpasswords inside logstash wrapper (or the other way around)?

**Comment by jakub-bochenski** [comment]
Verified that saving a project works w/o mask passwords installed.

I'm not sure if this will actually work during a build though (c.f. LogstashBuildWrapper.java) but I don't have a staging setup to check this now

**Comment by mwinter69** [comment]
Regarding pipeline you already have to define the mask passwords inside the logstash block (with the new logstash step I implemented). The LogstashBuildWrapper is not working with pipeline. 
Another thing is the JobDSL, here one might also need to make sure that logstash is set before mask passwords.
As of now if some has configured a regex in maskpasswords it will not be working as we don't have implemented this feature. Also by adding another time the maskpasswords it will be filtered twice.

**Comment by mwinter69** [comment]
I tested it locally, with mask passwords installed and configured.

**Comment by jakub-bochenski** [comment]
Sorry I've only now noticed that you removed the offending code from LogstashBuildWrapper. 
I must've been looking at master branch by mistake

**Comment by jakub-bochenski** [comment]
LGTM but are we sure that sending data will work without mask passwords installed?

**Comment by oleg-nenashev** [comment]
> LGTM but are we sure that sending data will work without mask passwords installed?

IIUC it won't

**Comment by jakub-bochenski** [comment]
To clarify: obviously the passwords won't be masked in that case.

On the other hand if sending data will fail anyway what's the point of having mask-passwords optional if the plugin will break the build without it?

**Comment by oleg-nenashev** [comment]
FYI @daniel-beck @Wadeck 

IMO we need a replacement for escaping engine so that sensitive data does not get exposed by default

**Comment by jakub-bochenski** [comment]
@oleg-nenashev I'm not sure. If one is already using plain text passwords visible in regular Jenikins console it's not that much worse if they are pushed to logstash in addition.

I think it's much more important to guard against a situation where mask-passwords in enabled, but is configured in wrong order. I think it could be possible to detect this and then add a warning badge etc.

**Comment by mwinter69** [comment]
Currently if you don't have mask passwords installed, you're not able to save freestyle jobs. And running them will also fail in case you directly push a config.xml with logstash enabled.
If you have it installed but haven't configured it in your job, the job will run and send passwords to logstash (if there are any).
@oleg-nenashev is it possible or a good idea via an ItemListener to make sure when a job is saved to reorder the buildwrappers if they are in the wrong order?

**Comment by mwinter69** [comment]
after looking at the sources of MaskPasswords I think using a BuildWrapper is the wrong approach for logstash. It should be ConsoleLogFilter.
The reason is that MaskPasswords also has a ConsoleLogFilter. If you configure centrally some passwords, they will be filtered regardless if you have MaskPasswords configured on a Project or not. But ConsoleLogFilters are initialized before the BuildWrapper which means logstash will send its data before it is filtered. Using a high ordinal should make sure that logstash is called before mask passwords.
This way we can remove the dependency to MaskPasswords and be sure that we always get filtered output

**Comment by mwinter69** [comment]
And by using a ConsoleLogFilter we could also enable it globally if desired.

File: src/main/java/jenkins/plugins/logstash/LogstashBuildWrapper.java [hunk 1]
```diff
@@ -24,100 +24,63 @@
...
-
-import org.kohsuke.stapler.DataBoundConstructor;
-
-import com.michelin.cio.hudson.plugins.maskpasswords.MaskPasswordsBuildWrapper;
-import com.michelin.cio.hudson.plugins.maskpasswords.MaskPasswordsBuildWrapper.VarPasswordPair;
-import com.michelin.cio.hudson.plugins.maskpasswords.MaskPasswordsConfig;
-
 /**
- * Build wrapper that decorates the build's logger to insert a
+ * Build wrapper that only indicates to the ConsoleLogFilter
```
Comment (jakub-bochenski) [inline src/main/java/jenkins/plugins/logstash/LogstashBuildWrapper.java]: I might be missing something but this looks like a noop to me -- I don't see any references to ConsoleLogFilter

Comment (jakub-bochenski) [reply src/main/java/jenkins/plugins/logstash/LogstashBuildWrapper.java]: nvm -- found the back-reference in ConsoleLogFilter

Can you reword the docs to state more explicitly that this is just a marker object and ConsoleLogFilter will check for it's presence?

File: src/main/java/jenkins/plugins/logstash/LogstashBuildWrapper.java [hunk 2]
```diff
@@ -24,100 +24,63 @@
...
+ * that it should forward log to an indexer. Keep this for backward compatibility.
+ * Better approach would be a JobProperty.
```
Comment (jakub-bochenski) [inline src/main/java/jenkins/plugins/logstash/LogstashBuildWrapper.java]: Can you please add a new issue to track this?

File: src/main/java/jenkins/plugins/logstash/LogstashBuildWrapper.java (hunk 1 above)
Comment (mwinter69) [reply src/main/java/jenkins/plugins/logstash/LogstashBuildWrapper.java]: done

File: src/main/java/jenkins/plugins/logstash/LogstashBuildWrapper.java (hunk 2 above)
Comment (mwinter69) [reply src/main/java/jenkins/plugins/logstash/LogstashBuildWrapper.java]: opened JENKINS-49960

**Pull request merged**